add a transformer suite in TP/SP tests

tianyu-l opened the pull request:
This is to address issue #115309.

Test plan
`python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_False`
`python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_True`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115530



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225

tianyu-l committed:
add a transformer suite in TP/SP tests

[ghstack-poisoned]

tianyu-l committed:
Update on "add a transformer suite in TP/SP tests"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l committed:
Update on "add a transformer suite in TP/SP tests"

This is to address issue #115309.

Test plan
`python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_False`
`python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_True`




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

wanchaol reviewed (commented):
Mostly looks good! have a few comments inlined, once those are addressed I think we are good to go!

tianyu-l committed:
Update on "add a transformer suite in TP/SP tests"

This is to address issue #115309.

Test plan
`python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_False`
`python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_True`




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l commented on test/distributed/tensor/parallel/test_tp_examples.py:
```diff
@@ -152,6 +284,10 @@ def test_mlp_inference(self):
+    @with_comms
+    @parametrize("is_seq_parallel", [True, False])
+    def test_transformer_training(self, is_seq_parallel):
+        self._test_transformer_training_e2e(is_seq_parallel=is_seq_parallel)
```
Removed. I was just following the style of the other two tests in this file.

wanchaol commented on test/distributed/tensor/parallel/test_tp_examples.py:
```diff
@@ -152,6 +158,147 @@ def test_mlp_inference(self):
+                    device_mesh,
+                    PrepareModuleInput(input_layouts=Shard(1), desired_input_layouts=Replicate())
+                )
+            parallelize_module(model_tp.layers[i].attention.wq, device_mesh, ColwiseParallel())
```
hmmm. I think we should have a separate parallelize_plan for each transformer layer defined as a dict of fqn -> parallel styles. Calling parallelize_module so many times is not preferred I think. 

Each layer we can have a parallelize_plan then a single `parallelize_module` call, and then the first two embedding layer and the final projection layer can call `paralllelize_module`.

tianyu-l replied:
Oh I see. I misunderstood your previous comment; in fact I also thought this way of calling is not good. I wonder in general what principles user should follow when parallelizing their model. E.g. should one always try to group submodules within a bigger module into a single parallelize_plan, as long as the plan does grow very big? What are the concerns of not having a super giant single plan, like what I did earlier?

wanchaol replied:
I think a super giant plan also works, it's just that when we have multiple layers with the same plan, things become a bit redundant, so I was thinking for transformer that have repeated sharding plan, per-layer sharding plan makes things more clean

tianyu-l committed:
Update on "add a transformer suite in TP/SP tests"

This is to address issue #115309.

Test plan
`python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_False`
`python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_True`




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm, thanks for working on this!

tianyu-l commented:
@pytorchbot merge

tianyu-l commented:
@pytorchbot merge
